common/src/api/internal/shared.rs

```diff
@@ -299,12 +299,81 @@ pub struct RouteConfig {
     pub vlan_id: Option<u16>,
 }
 
+#[derive(
+    Clone, Debug, Deserialize, Serialize, PartialEq, Eq, JsonSchema, Hash,
+)]
+pub struct UplinkAddressConfig {
+    pub address: IpNet,
+    /// The VLAN id (if any) associated with this address.
+    #[serde(default)]
+    pub vlan_id: Option<u16>,
+}
+
+impl UplinkAddressConfig {
+    pub fn addr(&self) -> IpAddr {
+        self.address.addr()
+    }
+}
+
+impl std::fmt::Display for UplinkAddressConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.vlan_id {
+            None => write!(f, "{}", self.address),
+            Some(v) => write!(f, "{};{}", self.address, v),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct UplinkAddressConfigError(String);
+
+impl std::fmt::Display for UplinkAddressConfigError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "parse switch location error: {}", self.0)
+    }
+}
+
+/// Convert a string into an UplinkAddressConfig.
+/// 192.168.1.1/24 => UplinkAddressConfig { 192.168.1.1/24, None }
+/// 192.168.1.1/24;200 => UplinkAddressConfig { 192.168.1.1/24, Some(200) }
+impl FromStr for UplinkAddressConfig {
+    type Err = UplinkAddressConfigError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let fields: Vec<&str> = s.split(';').collect();
+        let (address, vlan_id) = match fields.len() {
+            1 => Ok((fields[0], None)),
+            2 => Ok((fields[0], Some(fields[1]))),
+            _ => Err(UplinkAddressConfigError(format!(
+                "not a valid uplink address: {s}"
+            ))),
+        }?;
+        let address = address.parse().map_err(|_| {
+            UplinkAddressConfigError(format!(
+                "not a valid ip address: {address}"
+            ))
+        })?;
+        let vlan_id = match vlan_id {
+            None => Ok(None),
+            Some(v) => match v.parse() {
+                Err(_) => Err(format!("invalid vlan id: {v}")),
+                Ok(vlan_id) if vlan_id > 1 && vlan_id < 4096 => {
+                    Ok(Some(vlan_id))
+                }
+                Ok(vlan_id) => Err(format!("vlan id out of range: {vlan_id}")),
+            },
+        }
+        .map_err(|e| UplinkAddressConfigError(e))?;
+        Ok(UplinkAddressConfig { address, vlan_id })
+    }
+}
+
 #[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq, JsonSchema)]
 pub struct PortConfigV1 {
     /// The set of routes associated with this port.
     pub routes: Vec<RouteConfig>,
-    /// This port's addresses.
-    pub addresses: Vec<IpNet>,
+    /// This port's addresses and optional vlan IDs
+    pub addresses: Vec<UplinkAddressConfig>,
```

Comment:
I think this will be a breaking change for the bootstore, e.g. if we try to parse existing bootstore entries into this data structure, the serde deserialization will fail. We could take this on a few different ways.

1. Create a `PortConfigV2` and the associated conversion machinery.
2. Instead of changing the structure of `addresses`, add a `vlans` member that's a map indexed on address list which defaults to the empty set.

Reply:
Good catch.  I've tested install and reboot, but not install, reboot, upgrade, and reboot.

Reply:
Done.